src/Tools/CodeGenerator/Featurizers.yaml

```diff
@@ -639,6 +639,72 @@ featurizers:
       - input_type: TInputDouble
         output_type: double
 
+    # ----------------------------------------------------------------------
+  - name: MaxHorizonFeaturizer
+    estimator_name: MaxHorizonFeaturizer
+    release_version: 0.4.0
+
+    description: |-
+
+      This Featurizer will be completely implemented in Frameworks(ORT and ML.NET).
+
+      This Transformer adds rows according to maxHorizon and also appends a column
+      which is horizon_origin. The overall number of output rows is maxHorizon times
+      the input's number of rows.
+
+      The type of input and output is abstracted to Tabular which corresponding to
+      Tensor in OnnxRuntime and IDataView in ML.NET.
+
+      C++-style pseudo signature:
+        Tabular execute(Tabular const &value)
+
+      Examples:
+
+        The Transformer duplicate rows according to maxHorizon, and adding one
+        column called horizon_origin. The following example sets maxHorizon as 2.
+
+        Input:
+          +------------+-------+-------+
+          | date       | grain | sales |
+          +============+=======+=======+
+          | 2017-01-01 | A     |  0    |
+          +------------+-------+-------+
+          | 2017-02-01 | A     |  1    |
+          +------------+-------+-------+
+          | 2017-03-01 | A     |  2    |
+          +------------+-------+-------+
+          | 2017-04-01 | A     |  3    |
+          +------------+-------+-------+
+
+        Output:
+          +------------+-------+-------+----------------+
+          | date       | grain | sales | horizon_origin |
+          +============+=======+=======+================+
+          | 2017-01-01 | A     | 0     | 1              |
+          +------------+-------+-------+----------------+
+          | 2017-01-01 | A     | 0     | 2              |
+          +------------+-------+-------+----------------+
+          | 2017-02-01 | A     | 1     | 1              |
+          +------------+-------+-------+----------------+
+          | 2017-02-01 | A     | 1     | 2              |
+          +------------+-------+-------+----------------+
+          | 2017-03-01 | A     | 2     | 1              |
+          +------------+-------+-------+----------------+
+          | 2017-03-01 | A     | 2     | 2              |
+          +------------+-------+-------+----------------+
+          | 2017-04-01 | A     | 3     | 1              |
+          +------------+-------+-------+----------------+
+          | 2017-04-01 | A     | 3     | 2              |
+          +------------+-------+-------+----------------+
+
+    configuration_params:
+      - type: uint8

```

Comment:
>type: uint8 [](start = 8, length = 11)

I think we support a full spectre of types.

Reply:
I don't think so as this is a parameter that passing in from constructor

Reply:
Are you referring to a State to deserialze? Where do get that state from? This does not have any training part.

---
In reply to: [388535793](https://github.com/microsoft/FeaturizersLibrary/pull/159#discussion_r388535793) [](ancestors = 388535793)

Reply:
This is state that is created by the data scientist when constructing the model. It is not learned from the data, but rather a value that they provide.

A value of uint8 gives us a MaxHorizon value of 255; Is that enough for all scenarios? I think that the answer is probably yes, but it might be good to bump this up to a uint16 just to make sure.

Reply:
yeah this is only a transformer. what I mean is that(I am not very clear how ort does this, but I guess maxHorizon still serves as parameters, probably in ctx) maxHorizon is highly possible a limited and small number, do we really need to support full spectre for that?

Reply:
> This is state that is created by the data scientist when constructing the model. It is not learned from the data, but rather a value that they provide.
> 
> A value of uint8 gives us a MaxHorizon value of 255; Is that enough for all scenarios? I think that the answer is probably yes, but it might be good to bump this up to a uint16 just to make sure.

sure

Reply:
This would come from the attribute. I think it would be better if we just use more natural types for the CPU such as uint43 or int64

---
In reply to: [388542380](https://github.com/microsoft/FeaturizersLibrary/pull/159#discussion_r388542380) [](ancestors = 388542380)

Reply:
From the AutoML perspective, UINT32 should suffice.

Reply:
looks like uint32 is better